python-sdk/nuscenes/predict/models/mtp.py

```diff
@@ -0,0 +1,328 @@
+        :param frequency_in_hz: Frequency between timesteps in the prediction (in Hz).
+            Highest frequency is nuScenes is 2 Hz.
+        :param n_hidden_layers: Size of fully connected layer after the CNN
+            backbone processes the image.
+        :param input_shape: Shape of the input expected by the network.
+            This is needed because the size of the fully connected layer after
+            the backbone depends on the backbone and its version.
+
+        Note:
+            Although seconds and frequency_in_hz are labeled their
+            product should be an int.
+        """
+
+        super().__init__()
+
+        self.backbone = backbone
+        self.num_modes = num_modes
+        backbone_feature_dim = self._calculate_backbone_feature_dim(input_shape)
+        self.fc1 = nn.Linear(backbone_feature_dim + 3, n_hidden_layers)
+        predictions_per_mode = int(seconds * frequency_in_hz) * 2
+
+        self.fc2 = nn.Linear(n_hidden_layers, int(num_modes * predictions_per_mode + num_modes))
+
+    def _calculate_backbone_feature_dim(self, input_shape: Tuple[int, int, int]) -> int:
+        """ Helper to calculate the shape of the fully-connected regression layer. """
+        tensor = torch.ones(1, *input_shape)
+        output_feat = self.backbone.forward(tensor)
+        return output_feat.shape[-1]
+
+    def forward(self, image_tensor: torch.Tensor,
+                agent_state_vector: torch.Tensor) -> torch.Tensor:
+        """
+        Forward pass of the model.
+        """
+
+        backbone_features = self.backbone(image_tensor)
+
+        features = torch.cat([backbone_features, agent_state_vector], axis=1)
+
+        predictions = self.fc2(self.fc1(features))
+
+        # Normalize the probabilities to sum to 1 for inference.
+        mode_probabilities = predictions[:, -self.num_modes:].clone()
+        if not self.training:
+            mode_probabilities = f.softmax(mode_probabilities, dim=-1)
+
+        predictions = predictions[:, :-self.num_modes]
+
+        return torch.cat((predictions, mode_probabilities), 1)
+
+
+class MTPLoss:
+    """
+    Computes the loss for the MTP model.
+    """
+
+    def __init__(self,
+                 num_modes: int,
+                 regression_loss_weight: float,
+                 angle_threshold_degrees: float = 5.):
+        """
+        Inits MTP loss
+        :param num_modes: How many modes are being predicted for each agent.
+        :param regression_loss_weight: Coefficient applied to the regression loss to
+            balance classification and regression performance.
+        :param angle_threshold_degrees: Minimum angle needed between a predicted trajectory
+            and the ground to consider it a match.
+        """
+        self.num_modes = num_modes
+        self.num_location_coordinates_predicted = 2  # We predict x, y coordinates at each timestep.
+        self.regression_loss_weight = regression_loss_weight
```

Comment:
@freddyaboulton We've been using `1` as the default for the `regression_loss_weight` param for our models.

Reply:
Adding now!